### PR TITLE
[`Blip`] Remove redundant shift right

### LINKED ISSUE #23000: Possible bug in BlipForQuestionAnswering loss computation due to redundant right-shift
### System Info

- `transformers` version: 4.28.1
- Platform: Linux-5.15.0-69-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- Huggingface_hub version: 0.14.0
- Safetensors version: not installed
- PyTorch version (GPU?): 1.10.1+cu111 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: <fill in>
- Using distributed or parallel set-up in script?: <fill in>

### Who can help?

@ArthurZucker @younesbelkada 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

In `BlipForQuestionAnswering.forward()`, if the `labels` parameter is provided, then [lines 1218-1222](https://github.com/huggingface/transformers/blob/v4.28.1/src/transformers/models/blip/modeling_blip.py#L1218) set `decoder_input_ids` to the right-shifted version of `labels`, then passes both those variables into `self.text_decoder`, which is an instance of `BlipTextLMHeadModel`:

```python
if labels is not None and decoder_input_ids is None:
    # get decoder inputs from shifting lm labels to the right - this is used in training mode
    decoder_input_ids = self._shift_right(labels)
    # replace possible -100 values in labels by `pad_token_id`
    labels = labels.masked_fill(labels == self.decoder_pad_token_id, -100)

answer_output = self.text_decoder(
    input_ids=decoder_input_ids,
    attention_mask=decoder_attention_mask,
    encoder_hidden_states=question_embeds,
    encoder_attention_mask=attention_mask,
    labels=labels,
    return_dict=return_dict,
    reduction="mean",
)
```
However, in the code for `BlipTextLMHeadModel.forward()`, it seems like it's [already doing that shift for you](https://github.com/huggingface/transformers/blob/04ab5605fbb4ef207b10bf2772d88c53fc242e83/src/transformers/models/blip/modeling_blip_text.py#L888):
```python
if labels is not None:
  # we are doing next-token prediction; shift prediction scores and input ids by one
  shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
  labels = labels[:, 1:].contiguous().to(shifted_prediction_scores.device)
  loss_fct = CrossEntropyLoss(reduction=reduction, label_smoothing=0.1)
  lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
  if reduction == "none":
      lm_loss = lm_loss.view(prediction_scores.size(0), -1).sum(1)
```
Am I just misinterpreting this, or is the shift done twice, i.e., the loss is for next-next token prediction??

EDIT: As another point, the official [Jupyter notebook for BLIP](https://github.com/huggingface/notebooks/blob/main/examples/image_captioning_blip.ipynb) creates and instance of and trains `BlipForConditionalGeneration`, which also uses `BlipTextLMHeadModel` as the decoder. In this case, the `input_ids` and `labels` are the same (not shifted):
```python
for idx, batch in enumerate(train_dataloader):
  input_ids = batch.pop("input_ids").to(device)
  pixel_values = batch.pop("pixel_values").to(device)
  outputs = model(input_ids=input_ids,
                  pixel_values=pixel_values,
                  labels=input_ids)
```
Inside `BlipForConditionalGeneration.forward()`, it also doesn't shift the tokens:
```python
outputs = self.text_decoder(
            input_ids=input_ids,
            attention_mask=attention_mask,
            encoder_hidden_states=image_embeds,
            labels=labels,
            return_dict=return_dict,
            reduction="mean",
        )
```

EDIT 2: Seems like the original BLIP code similarly only shifts once. In `BLIP_VQA.forward()`, located [here](https://github.com/salesforce/BLIP/blob/main/models/blip_vqa.py#L51), there is no shift:
```python
answer = self.tokenizer(answer, padding='longest', return_tensors="pt").to(image.device) 
answer.input_ids[:,0] = self.tokenizer.bos_token_id
answer_targets = answer.input_ids.masked_fill(answer.input_ids == self.tokenizer.pad_token_id, -100)      

question_output = self.text_encoder(question.input_ids, 
                                    attention_mask = question.attention_mask, 
                                    encoder_hidden_states = image_embeds,
                                    encoder_attention_mask = image_atts,                             
                                    return_dict = True)    

question_states = []                
question_atts = []  
for b, n in enumerate(n):
    question_states += [question_output.last_hidden_state[b]]*n
    question_atts += [question.attention_mask[b]]*n                
question_states = torch.stack(question_states,0)    
question_atts = torch.stack(question_atts,0)     

answer_output = self.text_decoder(answer.input_ids, 
                                  attention_mask = answer.attention_mask, 
                                  encoder_hidden_states = question_states,
                                  encoder_attention_mask = question_atts,                  
                                  labels = answer_targets,
                                  return_dict = True,   
                                  reduction = 'none',
                                 )      
```
and there is a shift in `self.text_decoder.forward()`, as seen [here](https://github.com/salesforce/BLIP/blob/main/models/med.py#L904):
```python
prediction_scores = self.cls(sequence_output)
        
if return_logits:
    return prediction_scores[:, :-1, :].contiguous()  

lm_loss = None
if labels is not None:
    # we are doing next-token prediction; shift prediction scores and input ids by one
    shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
    labels = labels[:, 1:].contiguous()
    loss_fct = CrossEntropyLoss(reduction=reduction, label_smoothing=0.1) 
    lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
    if reduction=='none':
        lm_loss = lm_loss.view(prediction_scores.size(0),-1).sum(1)  
```

Only the `text_decoder` itself shifts the text (again, in the forward function).
### Expected behavior

N/A

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/23000

In fact `_shift_right` does not need to be called inside `BlipForQuestionAnswering` as the right shifting of the tokens is done directly on the text decoder as mentioned by the user. Therefore, that class will be trained to perform next-next token prediction instead of next-token prediction.

The fix is to simply remove that shift method and the call to it.

cc @sgugger

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
Thanks for fixing!
